services/src/modules/plugins/index.ts

```diff
@@ -0,0 +1,41 @@
+import * as fs from 'fs';
+import { join } from 'path';
+import { pluginsDir } from '../config';
+import logger from '../logger';
+import { StitchPlugin } from './types';
+
+const plugins: StitchPlugin[] = [];
+export let argumentInjectionGlobals: Record<string, unknown> = {};
+
+export async function loadPlugins() {
+  if (!pluginsDir) return;
+  const pluginList = await fs.promises.readdir(pluginsDir);
+
+  await Promise.all(
+    pluginList.map(async name => {
+      try {
+        const pluginImport = await import(join(pluginsDir!, name));
+        const plugin: Partial<Omit<StitchPlugin, 'name'>> =
+          typeof pluginImport === 'function' ? await pluginImport.call() : pluginImport;
+        plugins.push({ name, ...plugin });
+        logger.info({ name }, `Plugin ${name} has been loaded successfully.`);
+      } catch (err) {
+        logger.error({ err, name }, 'Failed to load plugin');
+        throw err;
+      }
+    })
+  );
+
+  try {
+    argumentInjectionGlobals = await buildArgumentInjectionGlobals();
+  } catch (err) {
+    logger.error({ err }, 'Failed to build argument injection globals');
+    throw err;
+  }
+}
+
+export async function buildArgumentInjectionGlobals(): Promise<Record<string, unknown>> {
+  return plugins
+    .filter(p => p.addArgumentInjectionGlobals)
+    .reduce(async (acc, cur) => ({ ...acc, ...(await cur.addArgumentInjectionGlobals!()) }), {});
```

Comment:
This can allow collisions with a later plugin overriding a value from a previous plugin.
We can try to avoid that by namespacing everything the plugin returns under the plugin name, not sure if we should though

Reply:
I think that name collision is the deployer problem. Adding namespaces will add a lot of verbosity. Private plugins' variables can be edited in case of collision and public plugins' variables should be well-named.